encoders/numeric/CompressionVaeEncoder/__init__.py

```diff
@@ -0,0 +1,76 @@
+                 *args, **kwargs):
+        """
+        :param model_path: specifies the path to the model, and also acts as the model name.
+        :param train_valid_split: controls the random splitting into train and test data. Here 99% of X is used
+        for training, and only 1% is reserved for validation.
+        :param dim_latent: specifies the dimensionality of the latent space.
+        :param iaf_flow_length: controls how many IAF(Inverse Autoregressive Flow) flow layers the model has.
+        :param cells_encoder: determines the number, as well as size of the encoders fully connected layers.
+        :param initializer: controls how the model weights are initialized, with options being `orthogonal` (default),
+        `truncated_normal`, and `lecun_normal`
+        :param batch_size: determine the batch sizes used for training.
+        :param batch_size_test: determine the batch sizes used for testing.
+        :param feature_normalization: tells CVAE whether it should internally apply feature normalization
+        (zero mean, unit variance, based on the training data) or not. If True, the normalisation factors are stored
+        with the model and get applied to any future data.
+        :param tb_logging: determines whether the model writes summaries for TensorBoard during the training process.
+        """
+        super().__init__(*args, **kwargs)
+        self.train_valid_split = train_valid_split
+        self.output_dim = output_dim
+        self.iaf_flow_length = iaf_flow_length
+        self.cells_encoder = cells_encoder
+        self.initializer = initializer
+        self.batch_size = batch_size
+        self.batch_size_test = batch_size_test
+        self.model_path = model_path
+        self.feature_normalization = feature_normalization
+        self.tb_logging = tb_logging
+
+    def post_init(self):
+        super().post_init()
+        self.to_device()
+
+    @batching
+    @as_ndarray
+    def encode(self, data: 'np.ndarray', *args, **kwargs) -> 'np.ndarray':
+        """
+        :param data: a `B x T` numpy ``ndarray``, `B` is the size of the batch
+        :return: a `B x D` numpy ``ndarray``
+        """
+        from cvae import cvae
+        model = cvae.CompressionVAE(data,
+                                    train_valid_split=self.train_valid_split,
+                                    dim_latent=self.output_dim,
+                                    iaf_flow_length=self.iaf_flow_length,
+                                    cells_encoder=self.cells_encoder,
+                                    initializer=self.initializer,
+                                    batch_size=self.batch_size,
+                                    batch_size_test=self.batch_size_test,
+                                    logdir=self.model_path,
+                                    feature_normalization=self.feature_normalization,
+                                    tb_logging=self.tb_logging)
+
```

Comment:
Is there a way to use a pretrained model? With this implementation a model will be trained for every batch.

Reply:
Yes, you are right. I think we have the same problem of the [TSNEEncoder](https://github.com/jina-ai/jina-hub/blob/master/encoders/numeric/TSNEEncoder/__init__.py) where the `fit_transform` method fits each batch into an embedded space and return that transformed output. 

Theoretically, we can use pre-trained models, but they don't exist. To generate them, we need to train this architecture with datasets that could make sense in this context.

What if we use [as_train_method](https://github.com/jina-ai/jina/blob/df7a7f494be7683a8377fdcf3bb8036ed7238e0a/jina/executors/decorators.py#L34) and [require_train](https://github.com/jina-ai/jina/blob/df7a7f494be7683a8377fdcf3bb8036ed7238e0a/jina/executors/decorators.py#L71) decorator functions to run the training one time before calling the encode function? or should I use a [PipelineEncoder](https://github.com/jina-ai/jina/blob/df7a7f494be7683a8377fdcf3bb8036ed7238e0a/jina/executors/encoders/__init__.py#L68)?

Reply:
The encoder for now must assume that it loads a pretrained model. It should not be training for now before loading. See as example CustomImageTorchEncoder, where a pretrained path is passed. TSNEEncoder is a spefial case because only fit transform is provided but because the algorithm is as such by design.

Reply:
Ok got it. I've updated the code, so now we can load a pretrained model. 

I'm wondering if in the future will make sense to create a specific TransformEncoder, that now is sklearn based, for other frameworks like Tensorflow, Keras and PyTorch.